### PR TITLE
ASoC step added

### PR DESCRIPTION
**Pull request by aboe026**

### CODE REVIEW

File: src/main/java/com/ibm/devops/connect/CRPipeline/UploadASoCTestResult.java [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    public String getRecordName() { return recordName; }
+    public String getCommitId() { return commitId; }
+
+    @Override
+    public void perform(final Run<?, ?> build, FilePath workspace, Launcher launcher, final TaskListener listener)
+            throws AbortException, InterruptedException, IOException {
+
+        Job parentJob = (Job)build.getParent();
+
+        Run firstBuild = parentJob.getBuildByNumber(1);
```
Comment (aboe026) [inline src/main/java/com/ibm/devops/connect/CRPipeline/UploadASoCTestResult.java]: why do we get the first build?

File: src/main/java/com/ibm/devops/connect/CRPipeline/UploadASoCTestResult.java [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+                    payload.put("record", record);
+                    payload.put("application", application);
+                    payload.put("build", buildObj);
+
+
+                    listener.getLogger().println("Uploading Payload Doc");
+                    try {
+                        CloudPublisher.uploadQualityDataRaw(payload.toString());
+                    } catch (Exception ex) {
+                        listener.error("Error uploading metric data: " + ex.getClass() + " - " + ex.getMessage());
```
Comment (aboe026) [inline src/main/java/com/ibm/devops/connect/CRPipeline/UploadASoCTestResult.java]: change to
```
Error uploading asoc data
```

File: src/main/java/com/ibm/devops/connect/CRPipeline/UploadASoCTestResult.java [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+                        build.setResult(Result.FAILURE);
+                    }
+
+                    listener.getLogger().println("UPLOADED PAYLOAD: " + payload.toString());
```
Comment (aboe026) [inline src/main/java/com/ibm/devops/connect/CRPipeline/UploadASoCTestResult.java]: I would put the payload printout before uploading. That way if there is an error, its easy to see what was attempted to be uploaded

File: src/main/java/com/ibm/devops/connect/CloudPublisher.java [hunk 4]
```diff
@@ -311,6 +316,38 @@ public static boolean uploadQualityData(HttpEntity entity) throws Exception {
         }
     }
 
+    public static void uploadQualityDataRaw(String payload) throws Exception {
+        CloudPublisher.ensureHttpClientInitialized();
+        String localLogPrefix= logPrefix + "uploadDeployment ";
```
Comment (aboe026) [inline src/main/java/com/ibm/devops/connect/CloudPublisher.java]: change to
```
uploadAsoc
```

File: src/main/java/com/ibm/devops/connect/CloudPublisher.java [hunk 5]
```diff
@@ -311,6 +316,38 @@ public static boolean uploadQualityData(HttpEntity entity) throws Exception {
...
+            HttpPost postMethod = new HttpPost(url);
+            attachHeaders(postMethod);
+            postMethod.setHeader("Content-Type", "application/json");
+            postMethod.setHeader("Authorization", "Bearer " + Jenkins.getInstance().getDescriptorByType(DevOpsGlobalConfiguration.class).getApiToken());
+            postMethod.setEntity(new StringEntity(payload));
+
+            response = httpClient.execute(postMethod);
+            resStr = EntityUtils.toString(response.getEntity());
+            if (response.getStatusLine().toString().contains("200")) {
+                log.info(localLogPrefix + "Uploaded Deployment successfully");
```
Comment (aboe026) [inline src/main/java/com/ibm/devops/connect/CloudPublisher.java]: change to
```
Uploaded Asoc successfully
```

File: src/main/java/com/ibm/devops/connect/CloudPublisher.java [hunk 6]
```diff
@@ -311,6 +316,38 @@ public static boolean uploadQualityData(HttpEntity entity) throws Exception {
...
+            } else {
+                throw new Exception("Bad response code when uploading Deployment: " + response.getStatusLine() + " - " + resStr);
```
Comment (aboe026) [inline src/main/java/com/ibm/devops/connect/CloudPublisher.java]: change to
```
Bad response code when uploading asoc
```

File: src/main/java/com/ibm/devops/connect/CloudPublisher.java [hunk 7]
```diff
@@ -311,6 +316,38 @@ public static boolean uploadQualityData(HttpEntity entity) throws Exception {
...
+            }
+        } finally {
+            if (response != null) {
+                try {
+                    response.close();
+                } catch (Exception e) {
+                    log.error("Could not close uploadDeployment response");
```
Comment (aboe026) [inline src/main/java/com/ibm/devops/connect/CloudPublisher.java]: change to
```
Could not close asoc response
```

File: src/main/java/com/ibm/devops/connect/CRPipeline/UploadASoCTestResult.java (hunk 1 above)
Comment (aberkIBM) [reply src/main/java/com/ibm/devops/connect/CRPipeline/UploadASoCTestResult.java]: Done

**Pull request merged**